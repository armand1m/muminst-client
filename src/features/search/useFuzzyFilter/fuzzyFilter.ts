import split from 'split-string';
import escapeRegExp from 'lodash.escaperegexp';
import { score } from 'fuzzaldrin';

interface Item {
  [key: string]: any;
}

/**
 * Filters a list of objects based on a given field
 *
 * fuzzyFilter filters objects based on the string stored in the "label" property.
 * The returned array will contain a subset of objects where the string
 * contains all search words (case-insensitive), sorted using a fuzzy matching
 * score. Quoted search substrings will filter based on the word ordering.
 *
 * @param {Object[]} items - A list of items to filter
 * @param {string} filterText - The string used to match objects in the array
 *
 * @example
 * const items = [
 *  {label: 'Pop Rocks'},
 *  {label: 'Charleston Chew'}
 * ];
 *
 * const results = fuzzyFilter(items, "rock");
 * // => [{label: 'Pop Rocks'}]
 */
export function fuzzyFilter<T extends Item>(
  key: string,
  items: T[],
  filterText: string
) {
  if (!filterText) {
    return items;
  }
  const splitOpts = { separator: ' ', quotes: ['"'] };
  const substrings = split(filterText, splitOpts).map((s) =>
    s.split('"').join('')
  );

  let result = substrings.reduce((filteredItems, substring) => {
    let regex: RegExp;
    try {
      regex = new RegExp(`^.*${substring}.*$`, 'i');
    } catch (e) {
      // Escape the substring only if necessary to enable filtering with
      // regular expressions
      const escapedSubstring = escapeRegExp(substring);
      regex = new RegExp(`^.*${escapedSubstring}.*$`, 'i');
    }
    return filteredItems.filter((item) => regex.test(item?.[key]));
  }, items);

  const getScore = (item: T) => score(item?.[key], filterText);
  result = result.sort((a, b) => getScore(b) - getScore(a));
  return result;
}
