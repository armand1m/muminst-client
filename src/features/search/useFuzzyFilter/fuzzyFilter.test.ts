import { fuzzyFilter } from './fuzzyFilter';

describe('fuzzyFilter(items, filterText)', () => {
  const items = [
    { label: 'New York' },
    { label: 'New Mexico' },
    { label: 'Alaska' },
  ];

  it.each`
    filterText
    ${''}
    ${'NEW'}
    ${'alas'}
    ${'new jersey'}
    ${'"new york"'}
    ${'"(me?x|al.ska+)"'}
  `(
    'should return the expected array if filterText equals "$filterText"',
    ({ filterText }) => {
      const result = fuzzyFilter('label', items, filterText);
      expect(result).toMatchSnapshot();
    }
  );
});
