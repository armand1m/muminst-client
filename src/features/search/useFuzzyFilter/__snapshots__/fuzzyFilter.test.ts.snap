// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fuzzyFilter(items, filterText) should return the expected array if filterText equals "" 1`] = `
Array [
  Object {
    "label": "New York",
  },
  Object {
    "label": "New Mexico",
  },
  Object {
    "label": "Alaska",
  },
]
`;

exports[`fuzzyFilter(items, filterText) should return the expected array if filterText equals ""(me?x|al.ska+)"" 1`] = `
Array [
  Object {
    "label": "New Mexico",
  },
  Object {
    "label": "Alaska",
  },
]
`;

exports[`fuzzyFilter(items, filterText) should return the expected array if filterText equals ""new york"" 1`] = `
Array [
  Object {
    "label": "New York",
  },
]
`;

exports[`fuzzyFilter(items, filterText) should return the expected array if filterText equals "NEW" 1`] = `
Array [
  Object {
    "label": "New York",
  },
  Object {
    "label": "New Mexico",
  },
]
`;

exports[`fuzzyFilter(items, filterText) should return the expected array if filterText equals "alas" 1`] = `
Array [
  Object {
    "label": "Alaska",
  },
]
`;

exports[`fuzzyFilter(items, filterText) should return the expected array if filterText equals "new jersey" 1`] = `Array []`;
